File: crack-data-structures-and-algorithms/leetcode/python-impl/reverse_linked_list_II_q92.py
# -*- coding: utf-8 -*-
# 0xCCCCCCCC


# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


def reverse_between(head, m, n):
    """
    :type head: ListNode
    :type m: int
    :type n: int
    :rtype: ListNode
    1 ≤ m ≤ n ≤ length of list.
    """
    if not head:
        return None

    # pp: pointer to prev
    # p: pointer to current
    # pn: pointer to next
    #       +-                -+
    # n1 -> |  n2 -> n3 -> n4  | -> n5
    # |     +- |     |        -+
    # pp       p     pn
    #
    #       +-                -+
    # n1 -> |  n3 -> n2 -> n4  | -> n5
    # |     +- |     |        -+
    # pp       pn    p
    #
    #       +-                -+
    # n1 -> |  n3 -> n2 -> n4  | -> n5
    # |     +-       |     |  -+
    # pp             p     pn
    #
    #       +-                -+
    # n1 -> |  n4 -> n3 -> n2  | -> n5
    # |     +- |           |  -+
    # pp       pn          p
    dummy = ListNode(0)
    dummy.next = head
    pp = dummy

    # range starts from 1
    pos = 1
    while pos < m:
        pp = pp.next
        pos += 1

    p = pp.next

    while pos < n:
        pn = p.next
        p.next = pn.next
        pn.next = pp.next
        pp.next = pn
        pos += 1

    return dummy.next

File: crack-data-structures-and-algorithms/leetcode/reverse_linked_list_II_q92.cpp

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// 核心思路
// 使用 extract-then-insert 处理框架
// 一开始令 p 指向第 m 个结点，然后每次操作，将 pn (pn = p.next) 结点抽出，插入 pp 结点后面
// 不断循环一直到处理完第 n 个结点
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if (!head) {
            return nullptr;
        }

        ListNode dummy(0);
        dummy.next = head;
        ListNode* pp = &dummy;

        int pos = 1;
        // Move forward m - pos steps to arrive at start
        for (; pos < m; ++pos) {
            pp = pp->next;
        }

        auto p = pp->next;

        // extract and insert
        for (; pos < n; ++pos) {
            auto pn = p->next;
            p->next = pn->next;
            pn->next = pp->next;
            pp->next = pn;
        }

        return dummy.next;
    }
};
